esi/src/lib.rs

```diff
@@ -23,210 +22,395 @@
+//!             Some(req),
+//!             // Use the default ESI configuration.
+//!             esi::Configuration::default()
+//!         );
 //!
-//!     // Execute the ESI document using the client request as context
-//!     // and sending all requests to the backend `origin_1`.
-//!     processor.execute_esi(req, beresp, &|req| {
-//!         Ok(req.with_ttl(120).send("origin_1")?)
-//!     })?;
+//!         processor.process_response(
+//!             // The ESI source document. Note that the body will be consumed.
+//!             &mut beresp,
+//!             // Optionally provide a template for the client response.
+//!             Some(Response::from_status(StatusCode::OK).with_content_type(mime::TEXT_HTML)),
+//!             // Provide logic for sending fragment requests, otherwise the hostname
+//!             // of the request URL will be used as the backend name.
+//!             Some(&|req| {
+//!                 println!("Sending request {} {}", req.get_method(), req.get_path());
+//!                 Ok(Some(req.with_ttl(120).send_async("mock-s3")?))
+//!             }),
+//!             // Optionally provide a method to process fragment responses before they
+//!             // are streamed to the client.
+//!             Some(&|req, resp| {
+//!                 println!(
+//!                     "Received response for {} {}",
+//!                     req.get_method(),
+//!                     req.get_path()
+//!                 );
+//!                 Ok(resp)
+//!             }),
+//!         )?;
+//!     } else {
+//!         // Otherwise, we can just return the response.
+//!         beresp.send_to_client();
+//!     }
 //!
 //!     Ok(())
 //! }
 //! ```
 
 mod config;
+mod document;
 mod error;
 mod parse;
 
-use fastly::http::body::StreamingBody;
-use fastly::http::header;
-use fastly::{Body, Request, Response};
-use log::{debug, error, warn};
+use fastly::http::request::PendingRequest;
+use fastly::http::{header, Method, StatusCode, Url};
+use fastly::{mime, Body, Request, Response};
+use log::{debug, error};
 use quick_xml::{Reader, Writer};
-use std::io::Write;
+use std::collections::VecDeque;
+use std::io::{BufRead, Write};
 
+pub use crate::document::Element;
 use crate::error::Result;
 pub use crate::parse::{parse_tags, Event, Tag};
 
 pub use crate::config::Configuration;
 pub use crate::error::ExecutionError;
 
 /// An instance of the ESI processor with a given configuration.
-#[derive(Default)]
 pub struct Processor {
+    // The original client request metadata, if any.
+    original_request_metadata: Option<Request>,
+    // The configuration for the processor.
     configuration: Configuration,
 }
 
 impl Processor {
-    /// Construct a new ESI processor with the given configuration.
-    pub fn new(configuration: Configuration) -> Self {
-        Self { configuration }
+    pub fn new(original_request_metadata: Option<Request>, configuration: Configuration) -> Self {
```

Comment:
I wonder if configuration should fully use the builder pattern here? So you'd replace `Processor::new` with `Configuration::build`. I don't feel strongly one way or the other but I'd lean slightly toward changing this.